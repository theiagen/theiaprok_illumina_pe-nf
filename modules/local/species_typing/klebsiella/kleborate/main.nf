process KLEBORATE {
    tag "$meta.id"
    label "process_medium"

    container 'us-docker.pkg.dev/general-theiagen/staphb/kleborate:2.2.0'

    input:
    tuple val(meta), path(assembly)

    output:
    tuple val(meta), path("*_kleborate_out.tsv")         , emit: kleborate_report
    tuple val(meta), path("SPECIES")                     , emit: kleborate_species
    tuple val(meta), path("MLST_SEQUENCE_TYPE")          , emit: kleborate_mlst_sequence_type
    tuple val(meta), path("VIRULENCE_SCORE")             , emit: kleborate_virulence_score
    tuple val(meta), path("RESISTANCE_SCORE")            , emit: kleborate_resistance_score
    tuple val(meta), path("NUM_RESISTANCE_GENES")        , emit: kleborate_num_resistance_genes
    tuple val(meta), path("BLA_RESISTANCE_GENES")        , emit: kleborate_bla_resistance_genes
    tuple val(meta), path("ESBL_RESISTANCE_GENES")       , emit: kleborate_esbl_resistance_genes
    tuple val(meta), path("KEY_RESISTANCE_GENES")        , emit: kleborate_key_resistance_genes
    tuple val(meta), path("GENOMIC_RESISTANCE_MUTATIONS"), emit: kleborate_genomic_resistance_mutations
    tuple val(meta), path("K_TYPE")                      , emit: kleborate_k_type
    tuple val(meta), path("K_LOCUS")                     , emit: kleborate_k_locus
    tuple val(meta), path("O_TYPE")                      , emit: kleborate_o_type
    tuple val(meta), path("O_LOCUS")                     , emit: kleborate_o_locus
    tuple val(meta), path("K_LOCUS_CONFIDENCE")          , emit: kleborate_k_locus_confidence
    tuple val(meta), path("O_LOCUS_CONFIDENCE")          , emit: kleborate_o_locus_confidence
    path "versions.yml"                                  , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    
    def skip_resistance = params.kleborate_skip_resistance ? '' : '--resistance'
    def skip_kaptive = params.kleborate_skip_kaptive ? '' : '--kaptive'
    def min_identity = params.kleborate_min_percent_identity ?: 90.0
    def min_coverage = params.kleborate_min_percent_coverage ?: 80.0
    def min_spurious_identity = params.kleborate_min_spurious_percent_identity ?: 80.0
    def min_spurious_coverage = params.kleborate_min_spurious_percent_coverage ?: 40.0
    def min_kaptive_confidence = params.kleborate_min_kaptive_confidence ?: 'Good'
    
    """
    # Run Kleborate on the input assembly
    kleborate \\
        ${skip_resistance} \\
        ${skip_kaptive} \\
        --min_identity ${min_identity} \\
        --min_coverage ${min_coverage} \\
        --min_spurious_identity ${min_spurious_identity} \\
        --min_spurious_coverage ${min_spurious_coverage} \\
        --min_kaptive_confidence ${min_kaptive_confidence} \\
        --outfile ${prefix}_kleborate_out.tsv \\
        --assemblies ${assembly} \\
        --all \\
        ${args}
    
    # Call parser
    kleborate_parser.py \\
        ${prefix}_kleborate_out.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        kleborate: \$(kleborate --version | sed 's/Kleborate v//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_kleborate_out.tsv
    touch SPECIES
    touch MLST_SEQUENCE_TYPE
    touch VIRULENCE_SCORE
    touch RESISTANCE_SCORE
    touch NUM_RESISTANCE_GENES
    touch BLA_RESISTANCE_GENES
    touch ESBL_RESISTANCE_GENES
    touch KEY_RESISTANCE_GENES
    touch GENOMIC_RESISTANCE_MUTATIONS
    touch K_TYPE
    touch K_LOCUS
    touch O_TYPE
    touch O_LOCUS
    touch K_LOCUS_CONFIDENCE
    touch O_LOCUS_CONFIDENCE

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        kleborate: \$(kleborate --version | sed 's/Kleborate v//' || echo "2.2.0")
    END_VERSIONS
    """
}