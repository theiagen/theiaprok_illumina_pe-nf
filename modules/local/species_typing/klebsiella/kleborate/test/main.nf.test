nextflow_process {
    name "Test Process KLEBORATE"
    script "../main.nf"
    process "KLEBORATE"

    test("KLEBORATE - test") {
        setup {
            run("GUNZIP") {
                script "../../../../../nf-core/gunzip/main.nf"
                process {
                    """
                    input[0] = [
                        [id:'test_sample'],
                        [file('https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/022/869/665/GCF_022869665.1_ASM2286966v1/GCF_022869665.1_ASM2286966v1_genomic.fna.gz', checkIfExists: true)]
                    ]
                    """
                }
            }
        }
        when {
            process {
                """
                input[0] = GUNZIP.out.gunzip
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.kleborate_bla_resistance_genes,
                    process.out.kleborate_k_locus,
                    process.out.kleborate_k_type,
                    process.out.versions
                ).match() }
            )
        }
    }

    test("KLEBORATE - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file(params.modules_testdata_base_path + 'genomics/prokaryotes/escherichia_coli/genome/genome.fa.gz', checkIfExists: true)
                ])
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}