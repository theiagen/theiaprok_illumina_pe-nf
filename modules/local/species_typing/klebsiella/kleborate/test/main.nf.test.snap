{
    "KLEBORATE - test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "BLA_RESISTANCE_GENES:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "K_LOCUS:md5,55c9a3d90a12d0f53306de998de72ca2"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "K_TYPE:md5,2c729497d91709de6bf1ccd875cf28e5"
                ]
            ],
            [
                "versions.yml:md5,316733758d57f5d0c42cffe9d5334953"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.5"
        },
        "timestamp": "2025-06-24T19:45:55.329932"
    },
    "KLEBORATE - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_kleborate_out.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "SPECIES:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "10": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "K_TYPE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "11": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "K_LOCUS:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "12": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "O_TYPE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "13": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "O_LOCUS:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "14": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "K_LOCUS_CONFIDENCE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "15": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "O_LOCUS_CONFIDENCE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "16": [
                    "versions.yml:md5,316733758d57f5d0c42cffe9d5334953"
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "MLST_SEQUENCE_TYPE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "VIRULENCE_SCORE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "RESISTANCE_SCORE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "NUM_RESISTANCE_GENES:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "BLA_RESISTANCE_GENES:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "ESBL_RESISTANCE_GENES:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "KEY_RESISTANCE_GENES:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "9": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "GENOMIC_RESISTANCE_MUTATIONS:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_bla_resistance_genes": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "BLA_RESISTANCE_GENES:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_esbl_resistance_genes": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "ESBL_RESISTANCE_GENES:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_genomic_resistance_mutations": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "GENOMIC_RESISTANCE_MUTATIONS:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_k_locus": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "K_LOCUS:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_k_locus_confidence": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "K_LOCUS_CONFIDENCE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_k_type": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "K_TYPE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_key_resistance_genes": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "KEY_RESISTANCE_GENES:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_mlst_sequence_type": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "MLST_SEQUENCE_TYPE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_num_resistance_genes": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "NUM_RESISTANCE_GENES:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_o_locus": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "O_LOCUS:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_o_locus_confidence": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "O_LOCUS_CONFIDENCE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_o_type": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "O_TYPE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_report": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample_kleborate_out.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_resistance_score": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "RESISTANCE_SCORE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_species": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "SPECIES:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "kleborate_virulence_score": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "VIRULENCE_SCORE:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,316733758d57f5d0c42cffe9d5334953"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.5"
        },
        "timestamp": "2025-06-24T19:45:59.709731"
    }
}